cmd/acraserver/listener.go

```diff
@@ -131,6 +208,54 @@ func (server *SServer) Start() {
 	}
 }
 
+func (server *SServer) Stop(socket *net.TCPListener) {
+	socket.SetDeadline(time.Now())
+}
+
+func (server *SServer) ListenerFD(socket *net.TCPListener) (uintptr, error) {
+	file, err := socket.File()
+	if err != nil {
+		return 0, err
+	}
+	return file.Fd(), nil
+}
+
+func (server *SServer) WaitAPI() {
+	server.cmAPI.Wait()
+}
+func (server *SServer) WaitACRA() {
+	server.cmACRA.Wait()
+}
+
+var WaitTimeoutError = errors.New("timeout")
+
+func (server *SServer) WaitWithTimeout(duration time.Duration) error {
+	timeout := time.NewTimer(duration)
+	wait := make(chan struct{})
+	go func() {
+		server.WaitACRA()
+		if server.config.GetWithZone() && server.config.GetEnableHTTPApi() {
+			server.Stop(server.socketAPI)
+		}
+		wait <- struct{}{}
+	}()
+
+	select {
+	case <-timeout.C:
+		return WaitTimeoutError
+	case <-wait:
+		return nil
+	}
+}
+
+func (server *SServer) Addr(socket *net.TCPListener) net.Addr {
```

Comment:
the same, we need support unix too

Reply:
Refactored

Reply:
done

Reply:
do we use or need this function? where it used?